### ISSUE TITLE
Amd Radeon Pro SSG Support

### ISSUE THREAD

**Issue by ryanwoods1706** [hashcat/hashcat#1487]
Would it be possible to implement the Amd Radeon Pro SSG api into hashcat, so people (who could afford it) will be able to crack passwords using GPUs. Although the # of people using this is limited it could draw more people to the program and encourage more people to try.

http://pro.radeon.com/_downloads/SSG_API_User_Manual%20v1.01.pdf

**Comment by magnumripper** [comment]
What use would a cracker have for that functionality? We're "able to crack passwords using GPUs" already, without vendor-specific stuff.

**Comment by ryanwoods1706** [comment]
Because when cracking an Ethereum wallet, depending on the wallets parameters you’re unable to crack it with a GPU, since it has huge memory requirements.

**Comment by Chick3nman** [comment]
Lets take a look at this thread for a bit and maybe it will clear a few things up.

https://hashcat.net/forum/thread-7170.html

TL;DR It does not seem like the SSG is usable in that way according to its documentation.

**Comment by ryanwoods1706** [comment]
Spoken to about 10/15 various hardware experts, and most have said it’ll work without any need of any implementation on the programs end, however some said that the program might need to use the API to intergrate with it. All have said it’s possible, it’s just whether it’ll work without the API implementation or not

**Comment by Chick3nman** [comment]
They don't understand how the SSG works then. The NVME drive(s) on the GPU are connected via the new HBCC(High Bandwidth Cache Controller) and act as a local file store for caching files that would normally be read from the host system storage, NOT as addressable mapped memory space. 

https://muropaketti.com/wp-content/uploads/2016/07/amd-capsaicin-radeon-pro-ssg-diagram-20160726.jpg

As you can see, the GPU Memory(VRAM) is separate from the NVME storage/cache.

And from the thread I linked above:

http://www.tomshardware.co.uk/amd-radeon-pro-ssg,news-53609.html

"AMD indicated that it is not using the NAND pool as a memory-mapped space currently..."

"The current implementation merely uses the SSD as a storage volume for frequently-accessed data..."

**Comment by ryanwoods1706** [comment]
I'm waiting to hear back from AMD about this, contacted them directly a few days ago so waiting for a reply. But why wouldn't it be possible to use the SSG as memory in a way meaning that it'll save stuff to files for when it needs it? Might be a bit of work but logically it seems possible

**Comment by ryanwoods1706** [comment]
https://www.pcworld.com/article/3099964/hardware/amds-new-ssg-technology-adds-an-ssd-to-its-gpu.html "The SSG memory would be treated as part of a huge pool of RAM. If the GPU can’t find its data in the local GDDR5+ or HBM RAM, it would then search the SSG. Only after that would it have to ask the CPU for what it wants." @Chick3nman

**Comment by Chick3nman** [comment]
From the article you linked:

"SSG would not replace graphics RAM itself, such as GDDR5+ or HBM or HBM2, as it would be too slow."

"When the GPU wants more work, it has to signal the CPU, which then fetches data from the larger pool of local system RAM or primary storage."

It's acting as a next step down from the VRAM, not as part of the local VRAM. Just like how Intel Optane is not RAM, it's simply a high speed cache that sits above the disks but below the system RAM. This is for when the GPU is done crunching its current work chunk and is going to get the next chunk of work from the disk. This does **NOT** allow you to increase the size of each work chunk beyond the size of the local HBM2 VRAM, it just lets you switch to the next chunk faster, increasing overall throughput. There seems to be some serious confusion here around the word "Memory". For the use case of acting as a Frame Buffer, the SSG will do very very well because it can load to/from the onboard SSD RAID very very quickly. But for our use case, we arent offloading a ton of file read/writes to a local cache. We are initializing a huge lookup table of random bit strings.

"The large memory requirements of scrypt come from a large vector of pseudorandom bit strings that are generated as part of the algorithm. Once the vector is generated, the elements of it are accessed in a pseudo-random order and combined to produce the derived key. A straightforward implementation would need to keep the entire vector in RAM so that it can be accessed as needed."

https://en.wikipedia.org/wiki/Scrypt

If we could split this RAM usage up magically into several pieces, some of which would go into the VRAM and the rest into system RAM/Disk efficiently, then the SSG would make sense. But we aren't doing that because doing so is NOT efficient.

Again, even if we wanted to get fancy with this and try to do that, the current SSG implementation does **NOT** allow you to do what you think it does, so we couldn't do it regardless.

"http://pro.radeon.com/_downloads/SSG_API_User_Manual%20v1.01.pdf

Pretty clear from the API documentation that the OpenCL extensions added are solely for file operations. And even if the API did have functions to use the onboard storage as VRAM (which, again, it does not), those functions would have to actually be implemented within Hashcat to make use of them, which would be a significant undertaking for a device that surely no one would use for password cracking. It wouldn't just magically work, and we would not be willing to invest the level of effort required to make it work. But again, it does not exist, so this is a non-issue. "

From the thread on the Hashcat forums that I linked earlier.

**Comment by ryanwoods1706** [comment]
Okay fair enough, but from what you’ve said that does that mean that if the GPU needs more memory it’ll go to the System and ask for some? (But it’ll be fucking slow)

**Comment by ryanwoods1706** [comment]
If we could split this RAM usage up magically into several pieces, some of which would go into the VRAM and the rest into system RAM/Disk efficiently, then the SSG would make sense. But we aren't doing that because doing so is NOT efficient.

From what you said there, would it be possible to do that with a normal GPU then? Like a 1080ti or something. Of course it wouldn’t be efficient, but it would surely increase hashrates (if possible) to maybe like anywhere from a couple hundred to 10,000? (Just a guess), but it just won’t be efficient

**Comment by Chick3nman** [comment]
It is currently not possible due to how the scrypt lookup table is being stored. If we found a way to write it into chunks and started storing internal states to the disk/sysRAM/etc. to get around the maximum VRAM limitation, it would very likely slow the whole process down even further. It would not make it more efficient. Scrypt is a Memory-Hardened algorithm and is very GPU resistant. It is designed specifically to make sure you can't do that efficiently. The ENTIRE lookup table needs to be accessible during computation since you need to look through following a pseudo-random order. Doing it in pieces would likely take more time loading/unloading than simply doing less of it concurrently but keeping the entire concurrent tables in the VRAM.

**Comment by ryanwoods1706** [comment]
So its literally impossible to do this on a GPU at all?

**Comment by Chick3nman** [comment]
GPUs can do it now, just not well due to the limitations of their VRAM. Which is _EXACTLY_ what SCrypt was designed to accomplish. It's high memory usage makes GPUs and other accelerated parallel processors require huge amounts of memory so that they arent faster/more efficient at cracking it. That was the goal of setting the SCrypt parameters that way when creating the wallet. To stop people from being able to crack it.

**Comment by ryanwoods1706** [comment]
Are there any parallel processors out there (that are good) cost isn’t an issue

**Comment by Chick3nman** [comment]
You would need to likely design an ASIC or use an FPGA with an absolutely ridiculous local memory size. It's simply not been done to my knowledge. You are far better off using many CPUs across many systems than designing your own processor. That will save you several million dollars.

**Comment by ryanwoods1706** [comment]
FPGA? Are there not ASIC scrypt machines out there? I’d jut need to load an OS on them and run it, and of course as a ton of RAM

**Comment by Chick3nman** [comment]
I do not believe you understand what an ASIC is or how they work. If you would like, I can explain it, but it might be better to do so on the forums. I will try to make this as simple as possible to avoid having to explain all of that.

Simple answer: No. There is nothing you can do except cluster a bunch of CPU machines together and hope for the best or deal with the speeds offered by your GPUs, if it runs on them at all. If you want, you can drop a couple million bucks and have a custom ASIC with a massive amount of memory designed to do your math, but most people don't have that kind of capital laying around to blow on something like this. There is no other magical way to make this work or suddenly make it faster. If there was, we wouldnt be using SCrypt for key derivation. Obviously.

**Comment by ryanwoods1706** [comment]
I've got a conference call with developers at AMD soon, but from the looks of our conversation so far it looks plausible:
“The SSG API is intended to allow you to use the 2TB of SSD storage as a cache layer for the GPU, so it should appear effectively as GPU memory. At this point we’re reaching the limits of what I can explain, and I would recommend getting a quick conference call scheduled between one of our developers and the customer – is that something you can arrange?”

Once I've had the call with AMD and we've reached a conclusion, if it is possible to use the 2TB of SSG in that way, would you (hashcat developers) be able to implement that into your software?

**Comment by ryanwoods1706** [comment]
@Chick3nman @jsteube

**Comment by Metatronovich** [comment]
> I've got a conference call with developers at AMD soon, but from the looks of our conversation so far it looks plausible:
> “The SSG API is intended to allow you to use the 2TB of SSD storage as a cache layer for the GPU, so it should appear effectively as GPU memory. At this point we’re reaching the limits of what I can explain, and I would recommend getting a quick conference call scheduled between one of our developers and the customer – is that something you can arrange?”
> 
> Once I've had the call with AMD and we've reached a conclusion, if it is possible to use the 2TB of SSG in that way, would you (hashcat developers) be able to implement that into your software?

Did you manage to gather more information on the subject?
I am very interested in the same thing.

**Comment by ryanwoods1706** [comment]
Nope, I ended up just spinning up a couple google cloud servers and cracking it that way. :P
